File: Solutions/Cpp/reverse_bits.cpp
/*
    [EASY]
    190. Reverse Bits

    Concepts:
    - bit manipulation

    Stats:
        Runtime | 0 ms      [Beats 100%]
        Memory  | 7.60 MB   [Beats 5.99%]
*/

class Solution {
public:
    uint32_t reverseBits(uint32_t n) {
        uint32_t num = 0;
        for (int i = 0; i < 32; i++)
            num |= (((n & 1 << (31-i)) >> (31-i)) << i);
        return num;
    }
};

File: Solutions/Python/reverse_bits.py
"""
    [EASY]
    190. Reverse Bits

    Concepts:
    - bit manipulation

    Stats:
        Runtime | 27 ms     [Beats 97.25%]
        Memory  | 17.20 MB  [Beats 8.20%]
"""

class Solution:
    def reverseBits(self, n: int) -> int:
        res = 0
        for i in range(32):
            # if ith digit set...
            if (n & (1 << i)):
                # set same bit on other half
                res |= (1 << (31 - i))
        return res